src/mimir_coordinator.py

```diff
@@ -81,49 +86,130 @@ def is_recommended(self) -> bool:
                 return False
         return True
 
-    def build_config(self, _charm_config: Dict[str, Any]) -> Dict[str, Any]:
+    def build_config(self, s3_data: _S3StorageBackend) -> Dict[str, Any]:
```

Comment:
again try to use variable names matching the type names:
if it's a 'backend', why are we calling it a 'data'?

either `s3_data: S3Data` or `s3_backend: S3Backend` (I suggest the former)